src/PhpDocumentor/ReflectionExtractor.php

```diff
@@ -24,11 +24,17 @@ public function extract(string $class): array
         $docBlockFactory = DocBlockFactory::createInstance(['inject' => Inject::class]);
         $classContext = (new ContextFactory())->createFromReflector($classReflection);
 
-        return \array_filter(\array_map(function (\ReflectionProperty $propertyReflection) use ($docBlockFactory, $classContext) {
+        $props = \array_filter(\array_map(function (\ReflectionProperty $propertyReflection) use ($docBlockFactory, $classContext) {
             $context = $this->getTraitContextIfExists($propertyReflection) ?? $classContext;
 
             return $this->createServiceProperty($propertyReflection, $docBlockFactory, $context);
         }, $classReflection->getProperties()));
+
+        if (false !== $parent = $classReflection->getParentClass()) {
+            return \array_merge($props, $this->extract($parent->getName()));
+        }
```

Comment:
What if a property is overridden in the child class?

Reply:
Indeed, problem arises when the overriding child property is annotated (as well as the parent) :thinking:
And what if the child property defines another type as `@var`or service id?

I would say that the rule should be that the extract method doesn't create another property once it's been defined on the child level.


```php
class A extends B
{
    /**
     * @inject bar.baz
     */
    private $prop1; // => contains @bar.baz 

    private $prop2; // => contains @Foo
}

class B
{
    /**
     * @inject foo.bar
     */
    private $prop1;

    /**
     * @inject Foo
     */
    private $prop2;
}
```

Reply:
Or maybe I'll just use traits...

Reply:
What if properties are used by methods in both `A` and `B`? We need to do more thinking 🤔 and testing 💥

Reply:
Here are examples to illustrate what I believe should be the correct behaviors.

```php
<?php

class A
{
    /** @inject 'Afoo' */
    public $foo; // contains 'Afoo'

    /** @inject 'Abar' */
    protected $bar; // contains 'Abar'

    /** @inject 'Abaz' */
    private $baz; // contains 'Abaz'

    public function __construct($foo, $bar, $baz)
    {
        $this->foo = $foo;
        $this->bar = $bar;
        $this->baz = $baz;
    }

    public function foo()
    {
        return $this->foo;
    }

    public function bar()
    {
        return $this->bar;
    }

    public function baz()
    {
        return $this->baz;
    }
}

class B extends A
{
    /** @inject 'Bfoo' */
    public $foo; // contains 'Bfoo'

    /** @inject 'Bbar' */
    protected $bar; // contains 'Bbar'

    /** @inject 'Bbaz' */
    private $baz; // contains 'Bbaz'
}

class C extends A
{
    // $this->foo contains 'Afoo'
    // $this->bar contains 'Abar'
    // No access to $this->baz, however $this->baz() returns 'Abaz'
}

class D extends A
{
    // $this->foo contains 'Afoo'
    // $this->bar contains 'Abar'

    private $baz; // contains null
}
```

As you mentionned, problem arises when accessing private properties through accessible methods in child class (illustrated in class C).

The lib already breaks encapsulation by using reflection anyway, you might want not to support this new feature at all if you want to keep it simple ^^.

Reply:
Shouldn't public/protected be shared? We could simply throw an exception if a property is redefined. Otherwise it should be safe to use.

Reply:
I agree, that should work (https://3v4l.org/cBorg) :thinking:

Reply:
It should be possible to override private property though (and inject only if it's been anotated).

I'm trying to write some tests to cover these cases.